### PR TITLE
feat(codex): plan and apply a Codex CLI update from bound evidence (#2811)

### PR DESCRIPTION
**Pull request by luvs01**
## Summary

Phase 2 of #2811, on top of the landed attestation slice (#4978). The `codex-cli-update` command grows the two mutating-path verbs the issue describes, while `check` and `attest` stay read-only:

- `ocx system codex-cli-update plan [--channel latest] [--json]` dry-runs an update: it resolves the exact registry version with its sha512 integrity, reads the process table fail-closed, and prints a plan id that digests the evidence the decision rests on. It writes nothing and installs nothing; a refusal is a normal dry-run answer and still exits 0.
- `ocx system codex-cli-update apply --plan <id> [--json]` installs exactly the version bound into the quoted plan id. The plan is recomputed from live evidence and refused unless the id still matches. It packs the resolved version, verifies the tarball sha512 against the plan-bound integrity, and installs only that verified file; never stops or restarts Codex, the app-server, the desktop app or the tray, classifies the outcome from a fresh inspection rather than the installer exit code, and repairs the shim only when this installation owned a matched shim before the update.

The parser is the authorization boundary: apply is unreachable without a well-formed plan id the operator read in a dry-run, and there is no plan state on disk to expire, collide or clean up. An unreadable process table refuses rather than reading as no live session.

Two installation boundaries are pinned down:

- Every npm call (`view`, `pack`, `install -g`) runs with `--registry=https://registry.npmjs.org`, an empty controlled `--userconfig`/`--globalconfig`, a sentinel-anchored cwd, and every `npm_config_*` env var stripped — a hostile project or user npm configuration cannot redirect the metadata, the integrity token, or the tarball. The resolved `dist.tarball` origin is validated against the pinned registry before a target can resolve.
- Apply acquires one cross-process update lease (`~/.opencodex/codex-cli-update.lock`, pid + payload, stale-pid/age reclamation) before the final process scan and holds it through the install and the readback. A second apply is refused `update_in_progress`, and the Codex startup paths this codebase controls — the remote-workspace app-server spawn and the desktop-app relaunch — observe the lease and wait or refuse rather than load a half-replaced install.

Regression coverage: `tests/codex-integration/codex-cli-update-plan.test.ts` exercises plan refusals, evidence drift, apply classification, registry isolation, and lease exclusion; `tests/cli/cli-codex-cli-update.test.ts` covers the merged check/attest/plan/apply grammar; `tests/clients/remote-workspace-codex-runtime.test.ts` and `tests/clients/desktop-app-restart.test.ts` cover startup observing the lease.

## Verification

Exact head: `d6fcdd5e7129838097322c26ec3348b590ab219d`, rebased onto dev `a499746395de97427182ac03917c248017094d1d`.

- `bun x tsc --noEmit`
- `bun test ./tests/cli/cli-codex-cli-update.test.ts`
- `bun test ./tests/codex-integration/codex-cli-update-plan.test.ts`
- `bun test ./tests/codex-integration/codex-cli-install-provenance.test.ts`
- `bun test ./tests/codex-integration/codex-cli-installation-identity.test.ts`
- `bun test ./tests/codex-integration/codex-cli-installation-targets.test.ts`
- `bun test ./tests/clients/remote-workspace-codex-runtime.test.ts`
- `bun test ./tests/clients/desktop-app-restart.test.ts`
- `bun test ./tests/cli/cli-registry.test.ts`
- `bun scripts/generate-ocx-skill-surface.ts --check`
- `bun scripts/structure-ssot.ts`
- `bun scripts/privacy-scan.ts`

## Review readiness checklist

This PR stays in draft until every box below is ticked. Tick all four boxes once the requirements are met:

- [x] All CI tests are green on my local testing.
- [x] I pushed my PR to the latest dev commit.
- [x] I resolved all correct Codex and CodeRabbit findings.

- [x] My PR is ready for review.

## New Features
- Added `codex-cli-update plan` to preview a version-pinned update.
- Added `codex-cli-update apply --plan <id>` to apply a validated update plan.
- Added JSON output and clearer status and refusal reporting for update operations.
- Added safeguards to verify installation evidence, process state, package integrity, and plan validity before updating.

## Bug Fixes
- Update operations now refuse stale plans, active sessions, unavailable process information, and ambiguous installation results without retrying or rolling back.

### CODE REVIEW

File: src/codex/cli-update-plan.ts [hunk 1]
```diff
@@ -0,0 +1,472 @@
...
+    return refusedApply(plan.refusal ?? "plan_unknown", plan);
+  }
+  // Any drift in ownership, installed version, location, target or shim eligibility
+  // changes the id. Refuse rather than regenerate: the operator would otherwise approve
+  // one plan and install another.
+  if (plan.planId !== planId) return refusedApply("plan_stale", plan);
+
+  const targetVersion = plan.targetVersion;
+  const before = plan.installedVersion;
+  const installer = (deps.runInstaller ?? defaultRunInstaller)(targetVersion);
```
[the remark this answers is not in the document]
Comment (luvs01) [reply src/codex/cli-update-plan.ts]: Fixed in 526ffaeb5. apply no longer runs npm install -g <pkg>@<version> directly: it npm-packs the planned version into a temp dir, recomputes the sha512 SRI of the fetched tarball, compares it with the plan-bound integrity, and installs only the verified local file. A digest mismatch now fails closed as integrity_mismatch before anything is written. Coverage: an artifact digest mismatch fails closed before install in codex-cli-update-plan.test.ts.

File: tests/codex-integration/codex-cli-update-zero-effect.test.ts [hunk 2]
```diff
@@ -195,7 +195,7 @@ describe("Codex CLI updater zero-effect boundary", () => {
     });
     expect(result.error).toBeUndefined();
     expect(result.status).toBe(2);
-    expect(result.stderr).toContain("codex-cli-update action must be check");
+    expect(result.stderr).toContain("codex-cli-update action must be check, plan or apply");
```
[the remark this answers is not in the document]
Comment (luvs01) [reply tests/codex-integration/codex-cli-update-zero-effect.test.ts]: Fixed in 526ffaeb5: the expected error string now reads check, attest, plan or apply to match the four-action usage.

**Comment by luvs01** [comment]
CodeRabbit follow-up handled at d1a1817:

- `src/cli/capabilities.ts:780` — the apply capability now describes the real flow: pack the resolved version, verify the tarball sha512 against the plan-bound integrity, install only that verified file.
- `src/codex/cli-update-plan.ts:266` — `defaultRunInstaller` fails closed (`integrityMismatch`) when `expectedIntegrity` is missing, so an injected caller cannot trigger an unverified install.
- `src/codex/cli-update-plan.ts:373` — the plan now compares semver precedence instead of raw strings: equal resolves to `already_current`, a lower target is refused as `target_not_newer`; a downgrade can no longer be applied.
- `src/codex/cli-update-plan.ts:389` (matched-shim lane) — intentional deferral, not dead code: a `matched` shim means the candidate path is the opencodex wrapper or its backing file, so ownership classification of that path would attest the shim rather than the real `@openai/codex` install. The inspector reports `shim_update_deferred` until the phase that follows `backingPath` and classifies the real install; `managed` is never true in this phase, so the applicable lane (including `shimEligible`) activates together when attestation is introduced. Keeping the lane in the contract now preserves `planId` semantics for that phase.

Local verification: `bun test` on the update-plan, version-skew, codex-cli-update and zero-effect suites (88 pass), `tsc --noEmit` clean; merged current dev (04761a188).

File: src/codex/cli-update-plan.ts [hunk 3]
```diff
@@ -0,0 +1,483 @@
...
+    // A failed readback, an unchanged version and a third version are all reported
+    // as-is. None of them is retried, and none of them is rolled back.
+    if (after !== null && after === before) return result("not_applied");
+    return result("ambiguous");
+  }
+
+  // A matched shim never reaches this point: the inspector reports a shim-owned
+  // candidate as standalone-unverified and the plan refuses `not_managed`, so an
+  // applied update never owned a shim npm could have replaced.
+  return result("applied");
```
[the remark this answers is not in the document]
Comment (luvs01) [reply src/codex/cli-update-plan.ts]: Valid — fixed in 228430bef. The post-install readback now requires readback.location === plan.location alongside the npm-global provenance and package-manifest checks, so a different npm-global candidate at the target version is classed ambiguous instead of applied. readback.managed was not added per the manifestOwned note. Regression case added to the ambiguous-readback test.

**Review by Ingwannu — CHANGES_REQUESTED** [review]
Requesting changes on exact head `99f22746585916f435be884d3c062ea5b64ed2bc`. The plan binding and tarball digest check are useful, but two installation-boundary blockers remain.

1. The integrity is not bound to the official registry. Both `npm view` and `npm pack` inherit the operator's/project's npm configuration. A local/project `.npmrc` can redirect `@openai/codex` to another registry; the version query, integrity query and downloaded tarball then agree with each other and the current sha512 check passes before the attacker's package lifecycle runs during global install. Pin the authoritative registry explicitly for metadata and pack, reject a returned tarball/source outside that origin, and isolate registry-affecting npm config while preserving only intentionally supported proxy/CA settings. Add a hostile project/user registry regression.

2. The process scan is a snapshot, not mutual exclusion. A Codex app-server can start after `createCodexCliUpdatePlan()` scans and before `npm install -g` mutates the global installation. Two apply commands can also pass the same plan concurrently and run global npm installs against the same prefix. Acquire one cross-process update lease before the final scan and hold it through install plus readback; Codex/service startup must observe the same lease and refuse/wait, or the claimed “never update during a live session” guarantee remains racy. Add start-after-scan and concurrent-apply regressions.

The plan/apply readback should stay fail-closed, and the verified tarball path must remain the only install input. This branch is also behind current `dev` and has no completed Cross-platform CI on the exact head; rebase and rerun after fixing these boundaries.

**Comment by luvs01** [comment]
@Ingwannu Both installation-boundary blockers are addressed on `d6fcdd5e7129838097322c26ec3348b590ab219d`, rebased onto dev `a499746395de97427182ac03917c248017094d1d`.

**1. Integrity is now bound to the official registry.** Every npm call (`view`, `pack`, `install -g`) runs with `--registry=https://registry.npmjs.org`, plus `--userconfig`/`--globalconfig` pointed at an empty controlled file, a cwd anchored by a sentinel `package.json` so a project `.npmrc` (or an ancestor walk) is never consulted, and every `npm_config_*` env var stripped — a scoped `@openai:registry` redirect cannot survive any channel. The intentionally supported proxy/CA settings (`HTTP(S)_PROXY`/`ALL_PROXY`/`NO_PROXY`, `NODE_EXTRA_CA_CERTS`) pass through untouched. The resolved `dist.tarball` URL is validated against the pinned origin before a target can resolve, so a digest-matching tarball hosted elsewhere is still refused. Regressions: hostile `npm_config_registry`/`npm_config_@openai:registry` env cannot reach the spawned npm; a tarball outside the origin is refused even with a valid digest.

**2. The process scan is now backed by one cross-process lease.** `apply` acquires `codex-cli-update.lock` (pid + payload, stale-pid/age reclamation, compare-and-delete release) before the plan recomputation — so the scan is mutual exclusion, not a snapshot — and holds it through `npm pack`, the sha512 check, `install -g`, and the readback. A concurrent apply on the same plan is refused `update_in_progress` before any evidence is gathered. Codex startup paths that observe the lease: the remote-workspace `codex app-server` spawn waits out a held lease then refuses, and the desktop-app relaunch skips with `update_in_progress`. Regressions: start-after-scan refusal, lease-clear-then-proceed, two concurrent applies, dead-holder reclaim, corrupt-lease recovery, release on every outcome.

Readback stays fail-closed and the verified tarball remains the only install input. Focused suites: 38/38 plan+lease+isolation tests, runtime and desktop-restart suites green, `tsc --noEmit` clean. Hosted Cross-platform CI on this head is `action_required` (fork approval gate) — it needs a maintainer-side run approval.

**Review by Ingwannu — CHANGES_REQUESTED** [review]
Re-reviewed exact head `d6fcdd5e7129838097322c26ec3348b590ab219d`.

The official-registry pinning fixes the prior mutable-registry blocker: metadata, pack, and install now use the pinned npm registry with project/user npm configuration isolated. The new cross-process lease is necessary, but its current stale/release protocol does not provide mutual exclusion under contention:

1. Two contenders can read the same stale record. Contender A unlinks it and creates its live lease; contender B then executes the unconditional stale `unlinkSync(path)`, deletes A's new lease, and acquires its own. Both installs can run.
2. An opener creates the lock file before writing the JSON record. A contender observing that short empty/truncated window treats it as corrupt, unlinks it, and acquires a second lease while the first owner is still writing.
3. Release compares only `ownerPid`. It has no unique token or file/directory identity, so a late release can unlink a successor associated with the same PID edge case.
4. A live owner is reaped solely because the record is older than 15 minutes. If an install/readback ever exceeds that estimate, mutual exclusion is deliberately broken rather than failed closed.

Use a self-identifying atomic-directory/token protocol (the repository's current service/shim locks are the relevant pattern), require same-token/path identity immediately before stale deletion and release, and never reap a live holder merely from age without an ownership-safe heartbeat/identity rule. Add deterministic two-contender tests for stale takeover, create-before-record publication, late release, and live-owner age boundaries.

The branch is also 14 commits behind current `dev`; rebase and rerun the explicit installation/security-boundary CI after fixing the lease.

**Comment by luvs01** [comment]
All four readiness items are now met: CI is green at head dd5fc5b, the branch is within 10 commits of dev, and every CodeRabbit finding has been resolved or answered (the latest readback-location fix in 228430bef was confirmed by CodeRabbit). The readiness checklist is ticked. Ready for review — the draft flag still needs a maintainer since ready-for-review is permission-gated for the fork author. cc @lidge-jun

**Comment by lidge-jun** [comment]
Closing without carrying (dropped in #5619's lane review). `plan` requires `managed: true`, but the production installation inspector never reports that. As a result, plan and apply only run against synthetic test input. Making them reachable needs a proof that the selected installation is managed, which is a separate design change, and the current `codex-cli-update.ts` edits also conflict with another in-flight change. The integrity-binding and CLI grammar work here would be welcome as a new PR built on that proof. Thank you for the thorough work!
